cmd/clair/introspection.go

```diff
@@ -0,0 +1,184 @@
+		fallthrough
+	case "prometheus":
+		endpoint := "/metrics"
+		if cfg.Metrics.Prometheus.Endpoint != nil {
+			endpoint = *cfg.Metrics.Prometheus.Endpoint
+		}
+		log.Info().
+			Str("endpoint", endpoint).
+			Msg("configuring prometheus")
+		promlog := log.With().
+			Str("component", "metrics-exporter").
+			Logger()
+		pipeline, hf, err := prometheus.InstallNewPipeline(prometheus.Config{
+			OnError: func(err error) {
+				promlog.Error().Err(err).Msg("prometheus error")
+			},
+		})
+		if err != nil {
+			return nil, err
+		}
+		srv.RegisterOnShutdown(pipeline.Stop)
+		mux.HandleFunc(endpoint, hf)
+	case "dogstatsd":
+		log.Info().
+			Str("endpoint", cfg.Metrics.Dogstatsd.URL).
+			Msg("configuring dogstatsd")
+		pipeline, err := dogstatsd.InstallNewPipeline(dogstatsd.Config{
+			URL: cfg.Metrics.Dogstatsd.URL,
+		})
+		if err != nil {
+			return nil, err
+		}
+		srv.RegisterOnShutdown(pipeline.Stop)
+	default:
+	}
+
+	// Trace config
+	traceCfg := sdktrace.Config{
+		// By default, assume an ingress is tagging incoming requests for
+		// tracing.
+		DefaultSampler: sdktrace.NeverSample(),
+	}
+	if cfg.Mode == config.DevMode {
+		// Unless we're in dev mode, then go ham.
+		traceCfg.DefaultSampler = sdktrace.AlwaysSample()
+	}
+	if p := cfg.Trace.Probability; p != nil {
+		traceCfg.DefaultSampler = sdktrace.ProbabilitySampler(*p)
+	}
+	traceOpts := []sdktrace.ProviderOption{
+		sdktrace.WithConfig(traceCfg),
+	}
+	if n := cfg.Trace.Name; n != "" {
+		log.Info().Str("sink", n).Msg("configuring trace sink")
+	}
+Trace:
+	switch cfg.Trace.Name {
+	case "stdout":
+		exporter, err := stdout.NewExporter(stdout.Options{})
+		if err != nil {
+			return nil, err
+		}
+		traceOpts = append(traceOpts, sdktrace.WithSyncer(exporter))
+	case "jaeger":
+		jcfg := &cfg.Trace.Jaeger
+		var e jaeger.EndpointOption
+		lev := log.Info()
+		switch {
+		case jcfg.Agent.Endpoint != "":
+			lev = lev.Str("agent", jcfg.Agent.Endpoint)
+			e = jaeger.WithAgentEndpoint(jcfg.Agent.Endpoint)
+		case jcfg.Collector.Endpoint != "":
+			var opt []jaeger.CollectorEndpointOption
+			u, p := jcfg.Collector.Username, jcfg.Collector.Password
+			if u != nil {
+				lev = lev.Str("username", *u)
+				opt = append(opt, jaeger.WithUsername(*u))
+			}
+			if p != nil {
+				lev = lev.Str("password", strings.Repeat("*", len(*p)))
+				opt = append(opt, jaeger.WithPassword(*p))
+			}
+			e = jaeger.WithCollectorEndpoint(jcfg.Collector.Endpoint, opt...)
+		default:
+			lev.Msg("neither jaeger collector nor agent specified")
+			break Trace
+		}
+
+		jaegerlog := log.With().
+			Str("component", "trace-exporter").
+			Logger()
+		opts := []jaeger.Option{
+			jaeger.WithOnError(func(err error) {
+				jaegerlog.Error().Err(err).Msg("jaeger error")
+			}),
+		}
+		if jcfg.BufferMax != 0 {
+			lev = lev.Int("buffer_max", jcfg.BufferMax)
+			opts = append(opts, jaeger.WithBufferMaxCount(jcfg.BufferMax))
+		}
+		if jcfg.ServiceName != "" || len(jcfg.Tags) != 0 {
```

Comment:
The service name is pretty much the only thing that is typically required when using Jaeger. As such, it would probably be a good idea to have a default service name in case it's not explicitly set.